### PR TITLE
Allow setting type for values in `__pydantic_extra__`

### PR DESCRIPTION
**Pull request by adriangb**
Selected Reviewer: @dmontagu

### CODE REVIEW

**Comment by adriangb** [comment]
Looks like there's some weirdness with the typing evaluation but I'll wait to fix that until I get some initial feedback on the proposal as a whole.

**Comment by adriangb** [comment]
please review

File: pydantic/_internal/_generate_schema.py [hunk 1]
```diff
@@ -259,6 +259,20 @@ def _model_schema(self, cls: type[BaseModel]) -> core_schema.CoreSchema:
...
 
+            extra_validator = None
+            if core_config.get('extra_fields_behavior', 'ignore') == 'allow':
+                for tp in (cls, *cls.__mro__):
+                    extras_annotation = cls.__annotations__.get('__pydantic_extra__', None)
+                    if extras_annotation is not None:
+                        tp = get_origin(extras_annotation)
+                        if tp not in (Dict, dict):
+                            raise PydanticSchemaGenerationError(
+                                'The type annotation for `__pydantic_extra__` must be a Dict[str, ...]'
```
Comment (dmontagu) [inline pydantic/_internal/_generate_schema.py:270]:
```suggestion
                                'The type annotation for `__pydantic_extra__` must be Dict[str, ...]'
```
minor, but reads better to me, feel free to ignore

File: pydantic/_internal/_generate_schema.py [hunk 2]
```diff
@@ -259,6 +259,20 @@ def _model_schema(self, cls: type[BaseModel]) -> core_schema.CoreSchema:
 
             model_validators = decorators.model_validators.values()
 
+            extra_validator = None
+            if core_config.get('extra_fields_behavior', 'ignore') == 'allow':
```
Comment (dmontagu) [inline pydantic/_internal/_generate_schema.py:263]:
```suggestion
            if core_config.get('extra_fields_behavior') == 'allow':
```
feel free to ignore

**Review by dmontagu — APPROVED** [review]
I would maybe suggest adding some docs before merging this as it's kind of a weird thing to document, would be nice to get the docs added as part of the feature. But ... approved ...

**Pull request merged**